File: src/services/domain-traffic.ts
import { invoke } from '@tauri-apps/api/core'

export type DomainTrafficRange = 'day' | 'week' | 'month' | 'year'

export interface DomainTrafficItem {
  domain: string
  proxyUpload: number
  proxyDownload: number
  directUpload: number
  directDownload: number
  proxyTotal: number
  directTotal: number
  total: number
}

export interface DomainTrafficTotals {
  proxyUpload: number
  proxyDownload: number
  directUpload: number
  directDownload: number
  proxyTotal: number
  directTotal: number
  total: number
}

export interface DomainTrafficStats {
  range: DomainTrafficRange
  fromDay: string
  toDay: string
  totals: DomainTrafficTotals
  items: DomainTrafficItem[]
}

export interface ClearDomainTrafficResult {
  removedDays: number
}

export async function getDomainTrafficStats(range: DomainTrafficRange) {
  return invoke<DomainTrafficStats>('get_domain_traffic_stats', { range })
}

/** Delete day buckets strictly before `beforeDay` (`YYYY-MM-DD`). */
export async function clearDomainTrafficBefore(beforeDay: string) {
  return invoke<ClearDomainTrafficResult>('clear_domain_traffic_before', {
    beforeDay,
  })
}

export async function clearDomainTrafficStats() {
  return invoke<ClearDomainTrafficResult>('clear_domain_traffic_stats')
}

File: src-tauri/src/cmd/domain_traffic.rs
use super::{CmdResult, StringifyErr as _};
use crate::module::domain_traffic::{
    ClearDomainTrafficResult, DomainTrafficManager, DomainTrafficRange, DomainTrafficStats,
};

#[tauri::command]
pub async fn get_domain_traffic_stats(range: DomainTrafficRange) -> CmdResult<DomainTrafficStats> {
    Ok(DomainTrafficManager::global().stats(range))
}

/// Delete day buckets strictly before `before_day` (`YYYY-MM-DD`).
#[tauri::command]
pub async fn clear_domain_traffic_before(before_day: String) -> CmdResult<ClearDomainTrafficResult> {
    DomainTrafficManager::global()
        .clear_before(&before_day)
        .await
        .stringify_err()
}

#[tauri::command]
pub async fn clear_domain_traffic_stats() -> CmdResult<ClearDomainTrafficResult> {
    DomainTrafficManager::global().clear_all().await.stringify_err()
}
